pkg/csi/service/wcp/controller.go

```diff
@@ -1040,12 +1046,115 @@ func (c *controller) ValidateVolumeCapabilities(ctx context.Context, req *csi.Va
 	}, nil
 }
 
+// ListVolumes returns the mapping of the volumes and corresponding published nodes.
 func (c *controller) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 	*csi.ListVolumesResponse, error) {
+	start := time.Now()
+	volumeType := prometheus.PrometheusBlockVolumeType
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
-	log.Infof("ListVolumes: called with args %+v", *req)
-	return nil, status.Error(codes.Unimplemented, "")
+	var err error
+	if !commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ListVolumes) {
+		return nil, status.Error(codes.Unimplemented, "")
+	}
+	controllerListVolumeInternal := func() (*csi.ListVolumesResponse, string, error) {
+		log.Debugf("ListVolumes called with args %+v, expectedStartingIndex %v", *req, expectedStartingIndex)
+		k8sVolumeIDs := commonco.ContainerOrchestratorUtility.GetAllVolumes()
+
+		startingIdx := 0
+		if req.StartingToken != "" {
+			startingIdx, err = strconv.Atoi(req.StartingToken)
+			if err != nil {
+				log.Errorf("Unable to convert startingToken from string to int err=%v", err)
+				return nil, csifault.CSIInvalidArgumentFault, status.Error(codes.InvalidArgument,
+					"startingToken not a valid integer")
+			}
+		}
+
+		// If the startingIdx is zero or not equal to the expectedStartingIndex, then
+		// it means the listVolume request is a new one and not part of a previous
+		// request, so fetch the volumes from CNS again
+		if startingIdx == 0 || startingIdx != expectedStartingIndex {
+			queryFilter := cnstypes.CnsQueryFilter{
+				ContainerClusterIds: []string{
+					c.manager.CnsConfig.Global.ClusterID,
+				},
+			}
+			querySelection := cnstypes.CnsQuerySelection{
+				Names: []string{
+					string(cnstypes.QuerySelectionNameTypeVolumeType),
+				},
+			}
+
+			cnsQueryVolumes, err := c.manager.VolumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+			if err != nil {
+				log.Errorf("Error while querying volumes from CNS %v", err)
+				return nil, csifault.CSIInternalFault, status.Error(codes.Internal, "Error while querying volumes from CNS")
+			}
+
+			cnsVolumeIDs = nil
+			for _, cnsVolume := range cnsQueryVolumes.Volumes {
+				cnsVolumeIDs = append(cnsVolumeIDs, cnsVolume.VolumeId.Id)
+			}
+		}
+
+		// If the difference between the volumes reported by Kubernetes and CNS
+		// is greater than the listVolumeThreshold, it might mean that the CNS
+		// cache is stale. So, fail the request and return an error
+		listVolumeThreshold := c.manager.VcenterConfig.ListVolumeThreshold
+		if len(k8sVolumeIDs)-len(cnsVolumeIDs) > listVolumeThreshold {
+			log.Errorf("Kubernetes and CNS volumes completely out of sync %v-%v=%v",
```

Comment:
`Kubernetes and CNS volumes completely out of sync` -> `Kubernetes and CNS volumes completely out of sync and exceeds the threshold: %d`

Also use %v to %d  as these are integers.